### ISSUE TITLE
How to see HTTP2 multiplexing/pipelining in action with Envoy forward proxy?

### ISSUE THREAD

**Issue by mamazik** [envoyproxy/envoy#24702]
# Background
Hello!

Can you please help me with the following issue? I'm quite new to Envoy, and to the forward proxy technology in general, so maybe I'm just missing something, but here goes nothing!

I'm using Envoy as dynamic HTTP forward proxy with the config below (should be very similar to what you provide at [this ](https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_filters/dynamic_forward_proxy_filter#config-http-filters-dynamic-forward-proxy) page):
```yaml
static_resources:
  listeners:
    - name: listener_0
      address:
        socket_address:
          protocol: TCP
          address: 127.0.0.1
          port_value: 10000
      filter_chains:
        - filters:
            - name: envoy.filters.network.http_connection_manager
              typed_config:
                "@type": type.googleapis.com/envoy.extensions.filters.network.http_connection_manager.v3.HttpConnectionManager
                stat_prefix: ingress_http
                http_protocol_options:
                  accept_http_10: true
                access_log:
                - name: envoy.file_access_log
                  typed_config:
                    "@type": type.googleapis.com/envoy.extensions.access_loggers.file.v3.FileAccessLog
                    path: "/var/log/envoy/envoy_access.log"
                route_config:
                  name: local_route
                  virtual_hosts:
                    - name: local_service
                      domains: ["*"]
                      routes:
                        - match:
                            prefix: "/"
                          route:
                            cluster: dynamic_forward_proxy_cluster
                            timeout: 0s
                        - match:
                            connect_matcher:
                              {}
                          route:
                            cluster: dynamic_forward_proxy_cluster
                            timeout: 0s
                            upgrade_configs:
                              - upgrade_type: CONNECT
                                connect_config:
                                  {}
                http_filters:
                  - name: envoy.filters.http.dynamic_forward_proxy
                    typed_config:
                      "@type": type.googleapis.com/envoy.extensions.filters.http.dynamic_forward_proxy.v3.FilterConfig
                      dns_cache_config:
                        name: dynamic_forward_proxy_cache_config
                        dns_lookup_family: V4_ONLY
                  - name: envoy.filters.http.router
                    typed_config:
                      "@type": type.googleapis.com/envoy.extensions.filters.http.router.v3.Router
  clusters:
    - name: dynamic_forward_proxy_cluster
      connect_timeout: 60s
      lb_policy: CLUSTER_PROVIDED
      cluster_type:
        name: envoy.clusters.dynamic_forward_proxy
        typed_config:
          "@type": type.googleapis.com/envoy.extensions.clusters.dynamic_forward_proxy.v3.ClusterConfig
          dns_cache_config:
            name: dynamic_forward_proxy_cache_config
            dns_lookup_family: V4_ONLY
admin:
  access_log_path: /var/log/envoy/admin_access.log
  address:
    socket_address:
      protocol: TCP
      address: 0.0.0.0
      port_value: 9901
```
Recently I got a task to get advantage of the `connection pooling` feature. I didn't really know how it's functioning, so for test purposes I wanted to start with configuring just 1 connection to any upstream host, over which the proxy would multiplex any amount of requests (I'm probably going to need to limit the amount of requests in the future, but during the test phase it doesn't matter).

# Issue
When I test the connection pooling with the config above, I just can't see it in action.

The documentation clearly [says ](https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/upstream/connection_pooling#http-2) that "The HTTP/2 connection pool establishes as many connections as are needed to serve requests. **With no limits, this will be only a single connection**."

What this means for me is the following: if I do 2 HTTP2 cURL requests to the same upstream host with some delay in between, the second request will use the TCP connection established by the first one. At least this is how I understand HTTP pipelining/multiplexing.

# Tests
The following commands are all executed on the proxy machine itself.

First cURL request:
```console
curl -Lvx localhost:10000 https://microsoft.com
```

Second cURL request:
```console
sleep 0.5;  curl -Lvx localhost:10000 https://microsoft.com
```

How I check the result:
```console
sleep 0.6; ss -pi | grep envoy
```

The result itself:
```console
tcp   ESTAB 0      0                                                     127.0.0.1:ndmp       127.0.0.1:42712   users:(("envoy",pid=229990,fd=36))
tcp   ESTAB 0      517                                             192.168.100.1:39524   20.81.111.85:https   users:(("envoy",pid=229990,fd=35))
tcp   ESTAB 0      0                                                     127.0.0.1:ndmp       127.0.0.1:42708   users:(("envoy",pid=229990,fd=34))
tcp   ESTAB 0      0                                               192.168.100.1:39536   20.81.111.85:https   users:(("envoy",pid=229990,fd=37))
```

In case if you're wondering, both request have this in the output, meaning they both use HTTP2 to connect to the endpoint:
```console
> GET / HTTP/2
> Host: www.microsoft.com
> User-Agent: curl/7.61.1
> Accept: */*
```

# Expectations
What I expect to see is just 1 established TCP connection towards 20.81.111.85:443, instead of 2.

According to the documentation, I don't even need to make any changes in the config for this to happen, but in any case I did try to play (to no avail) with multiple options like `max_connections` and `max_requests` in the `circuit breakers`, and `max_concurrent_streams` in `http2_protocol_options`.

**Comment by mamazik** [comment]
Today I got a bit farther in understanding what's going on.

According to the [HTTP2 RFC](https://httpwg.org/specs/rfc7540.html), connection is a layer 7 HTTP over TCP connection, and one such connection can handle multiple HTTP requests.

This part is clearly working fine, and can be seen just by running this single command:
```console
curl -Lv https://www.example.com https://www.example.com
```
(The second request will re-use the connection)

From what I understand this is as far as the protocol (HTTP2) goes, as in if you run a second cURL command, it's not going to reuse the connection of the first one.

However, if I understand it correctly, Envoy has its own layer of connection pooling, which sits on top of the protocol, and is totally agnostic to what the protocol does or doesn't (according to the documentation at least). Can you please confirm if that's true, and if yes, how this can be achieved?

**Comment by snowp** [comment]
Envoy should be performing connection pooling by default without any extra configuration. Do you happen to be running Envoy with multiple workers? Connection pooling is per thread, so it might be that you're seeing the connection handled by different workers?

Does this reproduce with `--concurrency 1`?

**Comment by ohadvano** [comment]
Can you try adding the upstream cluster with the following config, to explicitly specify that the upstream should be HTTP/2?
```
    typed_extension_protocol_options:
      envoy.extensions.upstreams.http.v3.HttpProtocolOptions:
        "@type": type.googleapis.com/envoy.extensions.upstreams.http.v3.HttpProtocolOptions
        explicit_http_config:
          http2_protocol_options: {}
```

Additionally, as far as I understood, the multiplexing will only occur if the streams are opened simultaneously. Meaning that if you send the first message and wait for some time, Envoy might close the connection. By the time the second request arrives at the listener, the underlying TCP connection might exist anymore.

**Comment by mamazik** [comment]
> Envoy should be performing connection pooling by default without any extra configuration. Do you happen to be running Envoy with multiple workers? Connection pooling is per thread, so it might be that you're seeing the connection handled by different workers?
> 
> Does this reproduce with `--concurrency 1`?

Hello!

Happy New Year, and thank you for the answer!

Yes, I think my setup indeed implied running Envoy with multiple workers, because I was simply running it with the following parameters: `/bin/envoy -c /etc/envoy/basic.yaml --base-id 0`

But when I added `--concurrency 1`, nothing seems to have changed. I verified that by opening 5 terminals locally, just to run the same command at the same time on 4/5 of them (as suggested by the user **@ohadvano**):
```console
curl -Lvx localhost:10000 https://microsoft.com
```
And then I check the amount of connections on the 5th terminal with this command:
```console
sleep 0.1; ss -pi | grep envoy
```
This still showed me 4 unique TCP connections, even when I did the same test with `max_connection_pools` set to 1 in `circuit breakers`:
```console
tcp   ESTAB      0      0                                                     127.0.0.1:ndmp         127.0.0.1:50738   users:(("envoy",pid=317355,fd=35))
tcp   SYN-SENT      0      0                                               192.168.100.1:46630   104.78.177.250:https   users:(("envoy",pid=317355,fd=31))
tcp   ESTAB      0      0                                                     127.0.0.1:ndmp         127.0.0.1:50724   users:(("envoy",pid=317355,fd=29))
tcp   SYN-SENT      0      0                                               192.168.100.1:46646   104.78.177.250:https   users:(("envoy",pid=317355,fd=34))
tcp   SYN-SENT      0      0                                               192.168.100.1:46642   104.78.177.250:https   users:(("envoy",pid=317355,fd=32))
tcp   SYN-SENT      0      0                                               192.168.100.1:46654   104.78.177.250:https   users:(("envoy",pid=317355,fd=36))
tcp   ESTAB      0      0                                                     127.0.0.1:ndmp         127.0.0.1:50730   users:(("envoy",pid=317355,fd=30))
tcp   ESTAB      0      0                                                     127.0.0.1:ndmp         127.0.0.1:50732   users:(("envoy",pid=317355,fd=33))
```
Same issue is observed if I run just 2 cURL commands with 1 second delay in between.

Also, like it was suggested by the user , I tried to put this to the upstream cluster config, but it didn't affect the issue in any way:
```yaml
    typed_extension_protocol_options:
      envoy.extensions.upstreams.http.v3.HttpProtocolOptions:
        "@type": type.googleapis.com/envoy.extensions.upstreams.http.v3.HttpProtocolOptions
        explicit_http_config:
          http2_protocol_options: {}
```

**Comment by mamazik** [comment]
Hello!

I noticed recently that our Envoy was coming from Tetrate (so it was almost 2 years old), and thought that probably that was the root cause.

However, nothing changed when I installed Envoy 1.25.0 and performed the same tests.

Do you have any more ideas please?

**Comment by mamazik** [comment]
Hello!

Totally accidentally, but it seems that I've found a clue... 

Today I tried to run some more tests, but I didn't prepend the target domain with https://, so I ran the commands like this:
```console
curl -Lvx localhost:10000 www.example.com
```
What I've noticed after that is that it created a constant TCP/IP connection, which didn't go away once the request has been completed.

Still don't know why, but I'm not seeing the same behavior when I do prepend the domain like this:
```console
curl -Lvx localhost:10000 www.example.com
```
Anyway, I can conclude that the feature does work for me, albeit in a strange way.

By the way, I'm still not sure if it's something that was brought by 1.25.0, or if it was like that even back on 1.18.2
